src/shell/shell_cmds.cpp

```diff
@@ -256,21 +256,41 @@ void DOS_Shell::CMD_DELETE(char * args) {
 	dos.dta(save_dta);
 }
 
-void DOS_Shell::PrintHelpForCommands(const HELP_LIST requested_list)
+#define PRINT_HELP_CHECK_ROWS if (++rows_printed == nrows) { \
+				CMD_PAUSE(empty_string); \
+				rows_printed = 0; \
+			}
+
+void DOS_Shell::PrintHelpForCommands(HelpUtil::Filter req_filter)
 {
 	BIOS_NROWS; // macro creates 'nrows' queried from BIOS
+	nrows--;
 	int rows_printed = 0;
-	for (const auto &s : shell_cmds) {
-		if (requested_list == HELP_LIST::COMMON && !s.second.flags)
-			continue;
-
-		WriteOut("<\033[34;1m%-8s\033[0m> %s", s.first.c_str(),
-		         MSG_Get(s.second.help));
+	using namespace HelpUtil;
+	for (const auto &cat : {Category::DOSBOX, Category::FILE, Category::BATCH, Category::MISC}) {
+		bool category_started = false;
+		for (const auto &s : get_help_list()) {
+			if (req_filter == Filter::COMMON &&
+				s.second.filter != Filter::COMMON)
+				continue;
+			if (s.second.category != cat)
+				continue;
+			if (!category_started) {
+				auto header_pattern = convert_ansi_markup("[color=blue]%s[reset]\n");
+				WriteOut(header_pattern.c_str(), category_heading(cat));
+				category_started = true;
+				// A page break might be needed after writing the header
+				PRINT_HELP_CHECK_ROWS;
+			}
```

Comment:
Why do you need `category_started`? Can't you just move this into the first loop, so you print it once per category?

Reply:
Because there might not be anything to print in that category, specifically when showing the abridged help screen.

Reply:
> Because there might not be anything to print in that category, specifically when showing the abridged help screen.

Ok, that makes sense.